[profiler] fix chrome tracing for profiler run with cuda

rohan-varma opened the pull request:
Summary:
There was an error in
https://github.com/pytorch/pytorch/pull/30724/files that resulted in
`export_chrome_trace` generating invalid JSON. This only came up when the
profiler is run with `use_cuda=True` from what it looks like. In the future, we
should have tests that ensure we generate valid JSON because we no longer use
the json library.

See https://github.com/pytorch/pytorch/issues/33961 for details on the issue and repro.
Test Plan: Add UT to validate JSON.

Differential Revision: D20171428

rohan-varma commented:
I think the test is failing because of permission error that seems to only be in windows.

peterjc123 commented on test/test_autograd.py:
```diff
@@ -2580,6 +2581,32 @@ def gen_matrices(p):
+            torch.add(t1, t2)
+
+        with tempfile.NamedTemporaryFile() as f:
+            prof.export_chrome_trace(f.name)
```
To make this work on Windows, either specify `tempfile.NamedTemporaryFile(delete=False)` or pass the raw object to this function `prof.export_chrome_trace(f)`.

rgommers replied:
The existing test at https://github.com/pytorch/pytorch/blame/master/test/test_autograd.py#L2734 simply skips Windows. 

I'd suggest passing `f` here; not deleting tempdirs results in a mess over time.

rohan-varma replied:
To pass `f` we'd have to modify the API/implementation of `export_chrome_trace` quite a bit - e.g. it currently expects a str, and also opens the file - so we'd have to modify it so that if we pass in a file-object, we don't open or close. I'll send a follow up PR to do that since that change could require more discussion created -  https://github.com/pytorch/pytorch/issues/34086 to keep track of this.

ezyang commented:
Windows test needs to be fixed, but in general the fix itself is good.

rohan-varma committed:
[profiler] fix chrome tracing for profiler run with cuda (#33987)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/33987

There was an error in
https://github.com/pytorch/pytorch/pull/30724/files that resulted in
`export_chrome_trace` generating invalid JSON. This only came up when the
profiler is run with `use_cuda=True` from what it looks like. In the future, we
should have tests that ensure we generate valid JSON because we no longer use
the json library.

Test Plan: Add UT to validate JSON.

Differential Revision: D20171428

fbshipit-source-id: c4c88e5fd921c20bb52b5955749cb6a21f1ad8f0

rohan-varma commented:
Reverting this because it broke a py 3.5 test on master: https://app.circleci.com/jobs/github/pytorch/pytorch/4676929

It doesn't look like we run this test on PRs so I didn't catch this.
